packages/ipfs-message-port-client/README.md

```diff
@@ -0,0 +1,146 @@
+
+
+## Lead Maintainer <!-- omit in toc -->
+
+[Alex Potsides](https://github.com/achingbrain)
+
+## Table of Contentens <!-- omit in toc -->
+
+- [Install](#install)
+- [Usage](#usage)
+- [Notes on Performance](#notes-on-performance)
+- [Contribute](#contribute)
+- [License](#license)
+
+## Install
+
+```bash
+$ npm install --save ipfs-message-port-client
+```
+
+## Usage
+
+This client library works with IPFS node over the [message channel][] and assumes that IPFS node is provided via `ipfs-message-port-server` on the other end.
+
+It provides following API subset:
+
+- [`ipfs.dag`](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/DAG.md)
+- [`ipfs.block`](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/BLOCK.md)
+- [`ipfs.add`](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/FILES.md#ipfsadddata-options)
+- [`ipfs.cat`](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/FILES.md#ipfscatipfspath-options)
+- [`ipfs.files.stat`](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/FILES.md#ipfsfilesstatpath-options)
+
+Client can be instantiated from the [`MessagePort`][] instance. Primary goal of
+this library is to allow sharing a node across browsing contexts (tabs, iframes)
+and therefore most likely `ipfs-message-port-server` will be in the separate JS
+bundle and loaded in the [SharedWorker][].
+
+
+```js
+const IPFSClient = require('ipfs-message-port-client')
+// URL to the script containing ipfs-message-port-server.
+const IPFS_SERVER_URL = '/bundle/ipfs-worker.js'
+
+const main = async () => {
+  const worker = new SharedWorker(IPFS_SERVER_URL)
+  const ipfs = IPFSClient.from(worker.port)
+  const data = ipfs.cat('/ipfs/QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1n')
+
+  for await (const chunk of data) {
+    console.log(chunk)
+  }
+}
+```
+
+It is also possible to instantiate detached client, which can be attach it to
+the server later on. This is useful when server port is received via message
+from other JS context (e.g. iframe)
+
+> Note: Client will queue all API calls and only execute them once it is
+> attached (unless they timeout or are aborted in the meantime).
+
+```js
+const IPFSClient = require('ipfs-message-port-client')
+
+
+const ipfs = IPFSClient.detached()
+
+const main = async () => {
+  const data = ipfs.cat('/ipfs/QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1n')
+
+  for await (const chunk of data) {
+    console.log(chunk)
+  }
+}
+
+window.onload = main
+window.onmessage = ({ports}) => {
+  IPFSClient.attach(ports[0])
+}
+```
+
+### Notes on Performance
+
+Since client works with IPFS node over [message channel][] all the data passed
+is copied via [structured cloning algorithm][], which may lead to suboptimal
+results (especially with large binary data). In order to avoid unnecessary
+copying all API options have being extended with optional `transfer` property
+that can be supplied [Transferable][]s which will be used to move corresponding
```

Comment:
I think boolean would be too coarse, because:

- This would force all or nothing choice, while in practice very likely it would be some.
- Things that contain transferrables include CIDs, binary data, etc... Caller of this API may not **own** some of those references (as they were passed to them by other caller) so they won't be capable of making the decision (which likely will push toward safer side of always copy).
- `{transfer: true}` would require internally arbitrary depth traversal to collect all the transferables.
- I'm not fan of this API, but it follows approach used by web APIs like [`postMessage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage).

I would like to explore ways to improve this, because it's not ideal, but I would prefer to do that in the future.